File: src/e2e/rtmp_multiquality_test.go
//go:build e2e

package e2e

import (
	"fmt"
	"net/http"
	"path/filepath"
	"testing"
	"time"

	"Theatrum/constants"
)

func TestRTMP_MultiQualityTranscoding(t *testing.T) {
	liveStreamKey := "multi-secret"
	username := "multiquality"
	token := computeXORToken(liveStreamKey, username)

	ts := setupTestServer(t, TestConfig{
		Channels: fmt.Sprintf(`  "/premium/{username}":
    stream:
      type: live
      path: "live/{username}"
      live_stream_key: "%s"
      auth_token_template: "{username}"
      qualities:
        low:
          width: 320
          height: 240
          framerate: 15
          bitrate: "200k"
          codec: "libx264"
          audio:
            bitrate: "64k"
            codec: "aac"
      distribution:
        hls:
          segment_duration: 2
          window_size: 3
`, liveStreamKey),
	})

	// Start RTMP stream
	rtmpURL := fmt.Sprintf("rtmp://127.0.0.1:%d/premium/%s", ts.RTMPPort, username)
	streamToRTMP(t, rtmpURL, token, 25)

	// Multi-quality: expect master.m3u8 at stream root
	masterPath := filepath.Join(ts.DataDir, "live", username, constants.MasterPlaylist)
	waitForFile(t, masterPath, 20*time.Second)

	// Expect quality subdirectory with playlist
	lowPlaylistPath := filepath.Join(ts.DataDir, "live", username, "low", constants.SubPlaylist)
	waitForFile(t, lowPlaylistPath, 20*time.Second)

	// Expect at least one segment in the quality dir
	lowSegmentPath := filepath.Join(ts.DataDir, "live", username, "low", "segment_000.ts")
	waitForFile(t, lowSegmentPath, 25*time.Second)

	// Fetch master playlist via HTTP
	masterURL := fmt.Sprintf("http://%s/premium/%s/%s", ts.HTTPAddr, username, constants.MasterPlaylist)
	resp := waitForHTTP(t, masterURL, 20*time.Second)
	resp.Body.Close()

	status, body := httpGetBody(t, masterURL)
	if status != http.StatusOK {
		t.Fatalf("Expected 200 for master playlist, got %d", status)
	}

	m3u8 := parseM3U8(body)
	if !m3u8.IsMaster {
		t.Error("Expected master playlist")
	}
	if len(m3u8.StreamPlaylists) == 0 {
		t.Error("Expected at least one quality variant in master playlist")
	}

	// Fetch quality sub playlist
	lowPlaylistURL := fmt.Sprintf("http://%s/premium/%s/low/%s", ts.HTTPAddr, username, constants.SubPlaylist)
	status, body = httpGetBody(t, lowPlaylistURL)
	if status != http.StatusOK {
		t.Fatalf("Expected 200 for low quality playlist, got %d", status)
	}

	subM3U8 := parseM3U8(body)
	if subM3U8.SegmentCount == 0 {
		t.Error("Expected at least one segment in low quality playlist")
	}
}

File: src/test/java/com/fjourdren/theatrum/e2e/RtmpMultiQualityTest.java
package com.fjourdren.theatrum.e2e;

import com.fjourdren.theatrum.domain.constant.VideoConstants;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;
import java.time.Duration;

import static org.assertj.core.api.Assertions.assertThat;

/** Port of Go's {@code rtmp_multiquality_test.go}: live transcoding into a quality subdirectory. */
class RtmpMultiQualityTest {

    private static final String KEY = "multi-secret";

    private static final String CHANNELS = """
              "/premium/{username}":
                stream:
                  type: live
                  path: "live/{username}"
                  live_stream_key: "%s"
                  auth_token_template: "{username}"
                  qualities:
                    low:
                      width: 320
                      height: 240
                      framerate: 15
                      bitrate: "200k"
                      codec: "libx264"
                      audio:
                        bitrate: "64k"
                        codec: "aac"
                  distribution:
                    hls:
                      segment_duration: 2
                      window_size: 3
            """.formatted(KEY);

    @TempDir
    static Path tempDir;

    static TestServerSupport server;

    @BeforeAll
    static void startServer() {
        TestServerSupport.requireFfmpeg();
        server = TestServerSupport.start(tempDir, CHANNELS);
    }

    @AfterAll
    static void stopServer() {
        if (server != null) {
            server.close();
        }
    }

    @Test
    void writesAMasterPlaylistAndAQualitySubdirectory() {
        String username = "multiquality";
        Process publisher = server.publish("premium/" + username, TestServerSupport.xorToken(KEY, username), 20);

        Path streamDir = server.dataDir.resolve("live").resolve(username);
        TestServerSupport.waitForFile(streamDir.resolve(VideoConstants.MASTER_PLAYLIST), Duration.ofSeconds(30));
        TestServerSupport.waitForFile(streamDir.resolve("low").resolve(VideoConstants.SUB_PLAYLIST),
                Duration.ofSeconds(30));
        TestServerSupport.waitForFile(streamDir.resolve("low").resolve("segment_000.ts"), Duration.ofSeconds(30));

        String masterUrl = "/premium/%s/%s".formatted(username, VideoConstants.MASTER_PLAYLIST);
        server.waitForHttp(masterUrl, Duration.ofSeconds(20));

        var master = server.get(masterUrl);
        assertThat(master.status()).isEqualTo(200);
        var masterPlaylist = TestServerSupport.parseM3u8(master.body());
        assertThat(masterPlaylist.master()).isTrue();
        assertThat(masterPlaylist.streamPlaylists()).isNotEmpty();

        var low = server.get("/premium/%s/low/%s".formatted(username, VideoConstants.SUB_PLAYLIST));
        assertThat(low.status()).isEqualTo(200);
        assertThat(TestServerSupport.parseM3u8(low.body()).segmentCount()).isPositive();

        TestServerSupport.kill(publisher);
    }
}
